Automatically reprepare query on EncodeError

josevalim opened the pull request:
Closes #123

fishcakez commented on integration_test/cases/stream_test.exs:
```diff
@@ -72,6 +72,48 @@ defmodule StreamTest do
       ] = A.record(agent)
   end
 
+  test "stream reprapres query on encode error" do
```
typo

fishcakez commented on lib/db_connection.ex:
```diff
@@ -1071,9 +1087,11 @@ defmodule DBConnection do
     try do
       DBConnection.Query.encode(query, params, opts)
+    rescue
+      DBConnection.EncodeError -> :retry
```
For clarity could we rename `:retry` to be `:prepare`, as it is not a straight retry, we have to do some extra work (prepare). Also we should return the existing `meter`, so its `{:prepare, meter}` so that we create it once.

fishcakez commented on lib/db_connection.ex:
```diff
@@ -80,15 +80,19 @@ defmodule DBConnection do
   defstruct [:pool_mod, :pool_ref, :conn_mod, :conn_ref, :conn_mode]
 
+  defmodule EncodeError do
+    defexception [:message]
```
Do we want to support `:value` for the term that was failed to be encoded? For Ecto this is the wrong place to handle this though so don't think we need it.

josevalim commented:
Btw, this patch does not work for Postgrex as is. If we attempt to prepare
an already prepared query, it will error. So we need to do something,
either by augmenting the query protocol or by passing an option to make
this work.

However, even by doing those changes, this is not enough to fix Ecto tests
on db_connection (there are 2 and one is related to this, you can see on
CI).
-- 


*José Valimwww.plataformatec.com.br
<http://www.plataformatec.com.br/>Founder and Director of R&D*

josevalim commented:
So it seems this patch is wrong? We should not call parse again when
falling back from execute to prepare_execute?
-- 


*José Valimwww.plataformatec.com.br
<http://www.plataformatec.com.br/>Founder and Director of R&D*

fishcakez commented:
Correct don't need to parse when it's already been done.

On 2 Sep 2018 00:30, "José Valim" <notifications@github.com> wrote:

So it seems this patch is wrong? We should not call parse again when
falling back from execute to prepare_execute?

-- 


*José Valimwww.plataformatec.com.br
<http://www.plataformatec.com.br/>Founder and Director of R&D*

—
You are receiving this because you commented.

Reply to this email directly, view it on GitHub
<https://github.com/elixir-ecto/db_connection/pull/141#issuecomment-417911026>,
or mute the thread
<https://github.com/notifications/unsubscribe-auth/AB6JTYc6u6y4ddKbhZihUk2By5xSOjbvks5uW4krgaJpZM4WWKJZ>
.

josevalim commented:
I have fixed the typo, changed it so we don't parse twice and moved to `{:prepare, meter}` as return. It should be good to go. I will try it with Postgrex master son.
